### PR TITLE
table: Avoid encode KindMaxValue of string type

### LINKED ISSUE #31721: Invalid utf8mb4 character string: 'FA'
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
set tidb_enable_list_partition=on;
drop table if exists PK_LCP9280;
CREATE TABLE `PK_LCP9280` (
  `COL1` char(1) NOT NULL,
  `COL2` varchar(20) DEFAULT NULL,
  `COL4` datetime DEFAULT NULL,
  `COL3` bigint(20) DEFAULT NULL,
  `COL5` float DEFAULT NULL,
  PRIMARY KEY (`COL1`) /*T![clustered_index] NONCLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
PARTITION BY LIST COLUMNS(`COL1`)
(PARTITION `P0` VALUES IN ("曊","鎿","藺","缋","谢","曣","悫","嚋","走","膖"),
 PARTITION `P1` VALUES IN ("纅","軁","殐","槠","到","觳","邡","ý","忷","灦"),
 PARTITION `P2` VALUES IN ("箶","稺","酯","躦","","ȹ","曉","叨","瓧","駾"),
 PARTITION `P3` VALUES IN ("觞","秴","茣","鴛","剭","搅","媺","湪","廼","癈"),
 PARTITION `P4` VALUES IN ("萔","覦","偁","Ͳ","书","坅","憡","椪","堥","瑝"),
 PARTITION `P5` VALUES IN ("鎽","騚","鋦","企","疮","痤","离","趩","顫","颭"),
 PARTITION `P6` VALUES IN ("飥","坙","吳","秱","鵑","篚","鱳","髳","茗","侳"),
 PARTITION `P7` VALUES IN ("乀","喷","栙","蒻","卽","旧","贪","靐","銹","藜"),
 PARTITION `P8` VALUES IN ("鴝","佳","唓","斑","笄","谵","瑞","矺","鬷","儱"),
 PARTITION `P9` VALUES IN ("婔","掺","暒","灍","繉","鑒","瀏","葞","馋","炧"),
 PARTITION `P10` VALUES IN ("籎","蘶"));
insert into PK_LCP9280 values("軁", "岰迕堺智獤膛鑰驻庥扎描慓綱槔ᛁ壟瑻荢螾髃", "4946-11-28 18:34:55", -40,-4.75602e37);
select * from PK_LCP9280 partition(p1, p3) where col1 != "篚" or col1 != "籎";
```

### 2. What did you expect to see? (Required)
```
MySQL [test]> select * from PK_LCP9280 partition(p1, p3) where col1 != "篚" or col1 != "籎";
+------+--------------------------------------------------------------+---------------------+------+-------------+
| COL1 | COL2                                                         | COL4                | COL3 | COL5        |
+------+--------------------------------------------------------------+---------------------+------+-------------+
| 軁  | 岰迕堺智獤膛鑰驻庥扎描慓綱槔ᛁ壟瑻荢螾髃 | 4946-11-28 18:34:55 |  -40 | -4.75602e37 |
+------+--------------------------------------------------------------+---------------------+------+-------------+
1 row in set (0.01 sec)
```
### 3. What did you see instead (Required)
```
MySQL [test]> select * from PK_LCP9280 partition(p1, p3) where col1 != "篚" or col1 != "籎";
ERROR 1300 (HY000): Invalid utf8mb4 character string: 'FA'
```

### 4. What is your TiDB version? (Required)

```
Release Version: v5.4.0-alpha-634-g02c4179
Edition: Community
Git Commit Hash: 02c4179dd2fe284f2364b24cbfe794467fe9141b
Git Branch: release-5.4
UTC Build Time: 2022-01-17 01:01:46
GoVersion: go1.17.6
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #31721

Problem Summary:

### What is changed and how it works?

The max value of string is `0xFA` which is an invalid UTF8 string. And the `genKey` func will encode this string which is useless, see [link](https://github.com/pingcap/tidb/pull/31743/files#diff-b162f5fd0f5d187791658b65d19334a3b8c3762542818bb1379f5163c65f5ed1L844). Just reorganize the code, avoid to encode string which type is `KindMaxValue` and `KindMinNotNull`.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Avoid encode KindMaxValue of string type
```

### CODE REVIEW

**Comment by Defined2014** [comment]
/cc @tangenta @xiongjiwei 
PTAL

**Review by zimulala — COMMENTED** [review]
Please reply "What is changed and how it works?"  in the description.

**Comment by Defined2014** [comment]
> link

Done.

**Comment by Defined2014** [comment]
> Please reply "What is changed and how it works?" in the description.

Done. @zimulala

**Review by xiongjiwei — APPROVED** [review]

File: table/tables/partition.go [hunk 1]
```diff
@@ -831,23 +833,24 @@ func (lp *ForListColumnPruning) LocateRanges(sc *stmtctx.StatementContext, r *ra
...
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
 
-	if lp.ExprCol.GetType().EvalType() == types.ETString {
-		// for string type, values returned by GetMinValue and GetMaxValue are already encoded,
-		// so it's unnecessary to invoke genKey to encode them.
-		if r.LowVal[0].Kind() == types.KindMinNotNull {
-			lowKey = (&lowVal).GetBytes()
+	// for string type, values returned by GetMinValue and GetMaxValue are already encoded,
```
Comment (zimulala) [inline table/tables/partition.go:837]:
```suggestion
	// For string type, values returned by GetMinValue and GetMaxValue are already encoded,
```

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**